src/jit/optcse.cpp

```diff
@@ -220,74 +202,125 @@ Compiler::fgWalkResult Compiler::optPropagateNonCSE(GenTree** pTree, fgWalkData*
     return WALK_CONTINUE;
 }
 
-/*****************************************************************************
- *
- *  Helper passed to Compiler::fgWalkAllTreesPre() to unmark nested CSE's.
- */
+//------------------------------------------------------------------------
+// Compiler::optUnmarkCSEs
+//    Helper passed to Compiler::fgWalkAllTreesPre() to unmark nested CSE's.
+// 
+//    argument 'pTree' is a pointer to a GenTree*
+//    argument 'data' contains pCallbackData which needs to be cast from void*
+//       to our actual 'wbKeepList' a pointer to a GenTree*
+//
+//    Any nodes visited that are in the 'keepList' are kept and we
+//    return WALK_SKIP_SUBTREES for them
+//    For any node not in the 'keepList' we call optUnmarkCSE and either
+//    remove the node and decrement the LclVar ref counts
+//    or add the subtree to 'wbKeepList' using gtBuildCommaList
+//    and return WALK_SKIP_SUBTREES.
+//
 
 /* static */
 Compiler::fgWalkResult Compiler::optUnmarkCSEs(GenTree** pTree, fgWalkData* data)
 {
-    GenTree*  tree     = *pTree;
-    Compiler* comp     = data->compiler;
-    GenTree*  keepList = (GenTree*)(data->pCallbackData);
+    GenTree*  tree       = *pTree;
+    Compiler* comp       = data->compiler;
+    GenTree** wbKeepList = (GenTree**)(data->pCallbackData);
+    GenTree*  keepList   = nullptr;
 
-    // We may have a non-NULL side effect list that is being kept
+    // If we have a non-Null wbKeepList
     //
-    if (keepList)
+    if (wbKeepList)
```

Comment:
When is it ever ok for `wbKeepList` to be null?

Reply:
It used to be allowed in the call from fgRemoveStmt, but I removed that case

Reply:
I can change this to a noway_assert